Source/engine/render/clx_render.cpp

```diff
@@ -284,107 +284,182 @@ void AppendOutlineRowSolidRuns(uint8_t x, uint8_t w, OutlineRowSolidRuns &solidR
 	}
 }
 
+#include <queue>
+#include <vector>
+
 template <bool SkipColorIndexZero>
 void GetOutline(ClxSprite sprite, OutlinePixels &result) // NOLINT(readability-function-cognitive-complexity)

```

Comment:
The new outline code seems to be a lot more expensive than before. What problem is it solving?
Are the new outline pixels different from before?
I'd appreciate some screenshots to help me understand.

Reply:
There shouldn't be any garbage pixels in the existing outline already, though maybe there is a bug in the code.

At least that's what the original PR description says (https://github.com/diasurgical/DevilutionX/pull/7144, further improved in https://github.com/diasurgical/DevilutionX/pull/7145):
> The new algorithm only fills pixels that will be visible.

So it should already be possible to render the outline after the sprite without changing the outline rendering code?

Reply:
> There shouldn't be any garbage pixels in the existing outline already, though maybe there is a bug in the code.
> 
> At least that's what the original PR description says (#7144):
> 
> > The new algorithm only fills pixels that will be visible.
> 
> So it should already be possible to render the outline after the sprite without changing the outline rendering code?

If you rearrange the calls, you'll see for example that Cain's belt renders outline pixels on nearly all of his belt. Griswold has a few stray pixels around his belt as well.

Reply:
Here are the outline pixels in the current master:

<img width="1605" height="918" alt="image" src="https://github.com/user-attachments/assets/a4c4cb4e-251f-4b64-8cdf-2c1bf3f7ba5b" />

Reply:
I see, the garbage pixels are where the black non-shadow pixels are.

I think they will also be rendered as semi-transparent because they will be considered a shadow by the current code?

Reply:
Can be a bit hard to tell where the shadows are even as a human:
<img width="306" height="433" alt="image" src="https://github.com/user-attachments/assets/1d3619fe-296a-4c7d-b9a3-0b1da5a602cb" />

Reply:
Yeah that seems incredibly problematic

Reply:
The actual conversion code probably wouldn't be too complicated

1. Sprite -> surface (`RenderClxSprite`)
2. Extract shadow from the surface into a new shadow surface.
3. Convert both surfaces back to clx (`SurfaceToClx`)

The rendering might even get a bit faster as there will no longer be a branch when rendering

Reply:
We can probably use OpenCV's [`getAffineTransform`](https://docs.opencv.org/2.4/modules/imgproc/doc/geometric_transformations.html?highlight=getaffinetransform#getaffinetransform) or [`findHomography`](https://docs.opencv.org/2.4/modules/calib3d/doc/camera_calibration_and_3d_reconstruction.html?highlight=findhomography#Mat%20findHomography%28InputArray%20srcPoints,%20InputArray%20dstPoints,%20int%20method,%20double%20ransacReprojThreshold,%20OutputArray%20mask%29) to find the affine transform used (assuming that's how they did the shadows).

Reply:
I'm assuming the the idea here is to generate masks that we apply when loading? Then we can do some manual tweaking after they have been initially generated.